### DISCUSSION TITLE
Question about Few Things and Thank you for supporting this! <3

### DISCUSSION THREAD

**Discussion by ashfidable** [kagenash1/godot-behavior-tree#13] [Q&A]
So, I have been looking into different behavior trees and everything to jumpstart. This is probably the best I have seen so far and loving the constant support. I tried few behaviors with timer and it's pretty fun to see the flow right on the tree.
I will start with the questions:

- Why did you opt for **physics_frame** and **idle_frame** approach than _normal process_ and _physics_process_? Not going to lie, I am loving this and did learn a new thing. Are they even different from godot's own built in functions (idle = process, physics = physics_process)?

- I created this custom patrol behaviour (it's barebones right now) but I keep getting assertion failed because of "not running()". The code is below.

**go_to_position.gd**
```swift
extends BTLeaf

func _tick(agent: Node, blackboard: Blackboard) -> bool:
	var ag = agent as Agent
	ag.move_towards_position(Vector2(85, 190), blackboard.get_data('delta'))
	if ag.global_position.distance_to(Vector2(85, 190)) <= 50:
		return succeed()
	return running()
```
**agent.gd**
```swift
extends KinematicBody2D
class_name Agent

export (NodePath) var blackboard_path
export (float) var speed

onready var blackboard = get_node_or_null(blackboard_path)
onready var sprite: Sprite = $Sprite

var velocity: Vector2

func _ready():
	blackboard.set_data('position', $Node/Position2D.global_position)

func _physics_process(delta):
	blackboard.set_data('delta', delta)

func move_towards_position(target_position: Vector2, delta: float):
	print(target_position)
	var difference = (target_position - position).normalized()
	velocity = move_and_slide(difference * speed, Vector2.UP)
	velocity.normalized()
```
_I am pretty sure it's my mistake, if so.. I would love to know fix and few pointers based on the script. You have been massive help to my workflow, I would love to continue using this._

**Comment by kagenash1** [comment]
First of all, thanks for the kind words! I'm glad to see this tool is helping someone and I'm doing my best to improve it and address everything.

As for you first question, in the next version I'm going to switch to the _physics_process and _process functions, because what I did there doesn't make much sense :) 

About your problem, right now the way to leave a node running is not to call run(), which is called automatically at each tick(by the way, running() is to check whether it is running), but to yield() for something. 
In fact, since run() is called automatically, if you yield() the tree will automatically understand that the node is running and progressively suspend the execution of all the higher nodes for optimization.
Your leaf then must only return either succeed() or fail(). If you yield(), then the node must be running().

With that being said, here are a few alternatives to make this work:

In this case, we consider reaching the goal a success, and not reaching it a failure.

**go_to_position.gd**
```swift
extends BTLeaf

var ag: Agent # Useless optimization
var destination: Vector2 = Vector2(85, 190)

func _tick(agent: Node, blackboard: Blackboard) -> bool:
	ag = agent
	ag.move_towards_position(destination, blackboard.get_data('delta'))
	if ag.global_position.distance_to(destination) <= 50:
		return succeed()
        else:
                return fail()
```
EDIT: The above example allows you to do something cool. You can put this leaf under a "bt_repeat_until" and set it until success.

or, let's change the way we think of it a little bit. The action is successful if you move, not if you reach the goal.

**agent.gd**
```swift
func move_towards_position(target_position: Vector2, delta: float) -> Vector2:
	print(target_position)
	var difference = (target_position - position).normalized()
	velocity = move_and_slide(difference * speed, Vector2.UP)
	#velocity.normalized() # I'm not sure what the purpose of this is, it return a new Vector2 but it is not assigned to anything.
        return velocity
```
**go_to_position.gd**
```swift
extends BTLeaf

var ag: Agent # Useless optimization
var destination: Vector2 = Vector2(85, 190)

func _tick(agent: Node, blackboard: Blackboard) -> bool:
      ag = agent
      var velocity = ag.move_towards_position(destination, blackboard.get_data('delta'))
      blackboard.set_data("distance_to_destination", agent.global_position.distance_to(destination))

       if velocity.length() > 0.1: # If we moved
		return succeed()
       else:
                return fail()
```

So what? Now you know the distance to the destination, so after the agent moved (successfully) you can have a conditional node (a decorator) where you say if distance <= 50 execute whatever you have to execute. 

I think the last option makes the most sense, because why setting a leaf to a running state if it's not running? You are gonna call it every frame, so it is gonna run and complete successfully at each frame. The action is to "take one step", not to reach the goal, so it is successful at each step :)

Moreover, another reason why your leaf cannot be running, is that you are doing a move_and_slide() call at each frame. By definition, if the leaf is running(), it will not be ticked, because the tree suspends execution while it is running. This is wrong because you do wanna tick it at every frame, and thus you wanna either succeed() or fail() at each tick() call.

(I'm assuming you know how to use yield, but if you don't I suggest you to check the documentation in the GDScript Basics page, at 'Yield and Coroutines'. On top of that, I suggest you to check the two script templates I included, where you find an example on how to do all the things you might wanna do with leaves and decorators.)

This whole process is gonna change in 1.2.0, where I plan to use the run() method for consistency and to even allow aborting running states, but until then this is how it works. I will try to make the future update so that it will not break your previous trees too much.

I hope I clarified your doubts and if you need more support feel free to ask :) I'm open to suggestions and change proposals if they are requested!

**Comment by ashfidable** [comment]
First of all, I would like to thank you so much for replying as soon as possible. You are a life savior for real and yes I am aware of yields (pretty powerful coroutines really). So, I did read the whole thing and honestly the last one works pretty well to me due to separation of distance check and the whole move_towards_position thing. Happy to say, it's working great here.

_My bad about velocity.normalized() thing, it was from my main project and I had to cut out most of the code to paste it here for easy debugging - sadly forgot to take that off haha_

Honestly, I do like this idle_frame and physics_frame thing. It looks nicer but that is no surprise that many here are used to physics_process and process instead. However, I am happy either way and loving the fact that you are catering towards user experience more. 

Coming to think of it, how would you go about implementing patrol between x number of points?
I was trying it now and most probably my head isn't working due to sleep but here's how I structured it for now.

* _BT_
* _BTSequence_
  * _BTSequence_
    * **AssignPoints** (Assigning a point from waypoint_list to a blackboard key called current_point)
    * **GoToPosition** (Same thing as above but instead of hardcoded Vector2, its referring to current_point key)
    * **Distance Check** (Decorator)
      * **_AssignAnotherPoint_**? (But then this will get changed when it repeats the tree again, so it's pretty pointless here)

Why the awful suggestion for the second AssignAnotherPoint was because the first AssignPoint keeps on repeating itself, so the current point keeps on changing. Let me quickly type out basic code for choosing a point from list

**change_points.gd**
```swift
extends BTLeaf


export (Array, NodePath) var waypoint
var curr_point = 0


func _tick(agent: Node, blackboard: Blackboard) -> bool:
	curr_point = (curr_point + 1) % waypoint.size()
	blackboard.set_data('current_point', get_node(waypoint[curr_point]))
	return succeed()
```

_Also, is it weird to say I am hyped about 1.2.0 than Godot 4.0?_  - you are killing it with the updates. When do you think that's going to come and will it change fundamentally when supporting nodes? Although, that's not an issue. Logic anyways in every node stays the same for sure. So, it's hardly any problem haha.

**Comment by kagenash1** [comment]
The first thing I can suggest is to think of this in the opposite order, first check if it's reached, and move if it's not.
So basically:

_BTSelector (instead of sequence)
_ _ IsPointReachedOrNull? (extends BTDecorator) 
_ _ _ AssignPoint
_ _ Move

If you look carefully I said reached or "null". You have a lot of freedom on this, but in general you may want to have some additional information (in this case, when the point is null) to know when to start the whole process.

I've been developing this for months and sometimes I forget the reason why I did things. There reason why I didn't use process is because I wanted to stop execution, and process is called at every frame. But that can easily be fixed with set_process(false), which, incredibly, I didn't know of at the time lol.

The visual editor is just gonna be on a higher level and won't change the actual logic (it will dynamically create a SceneTree).
As for all the other changes and improvements, I will make sure they are backward compatible, since there are already a few people using this. 
So far I've been pushing changes recklessly cause I was the only one using it, but now I have to be more careful 😄 

I'm flattered that you are waiting for the update, ha! Godot 4 will bring lots of crazy improvements that I'm looking forward to to improve the Behavior Tree. But I have a whole LOT of improvements and future versions planned, as I'm actively using this in my game and sharing the improvements as I go.

**Reply by ashfidable** [reply]
Sorry about the discussion in Issue instead. I tried thinking differently like yours and it does make sense now. Added with couple of always failure decorator to force the selector to go to "GoToPosition" node all the way even after assigning. Just made the behaviors and all so much modular! Thank you for helping me with this.

Coming to think of it, I like the ability of idle_frame and physics_frame. Really gives the edge on how to go forward with each frame. I am liking this a lot. 

The visual editor might be interesting, knowing this package it will be a quality feature than just stacked on graphedit haha. 
If you don't mind me asking, you have used this in your projects quite extensively. Did you come to any performance hurdle yet? I am not worried at all, but would love to know your story on this. Behavior Trees can be bit taxing anyway, I tried with 30 duplicated enemies in my game (same btree) and it's constant 60fps throughout - though my btree isn't that complex yet.

**Reply by kagenash1** [reply]
I had quite complex trees (enemy  detecting you, shooting, reloading, taking cover, changing position when hit depending on surroundings and awareness of the player) and I tested it with hundreds of enemies and never lost a single fps :) 
Godot is very optimized in terms of the SceneTree so using that makes it very fast.

**Reply by ashfidable** [reply]
Pretty amazing really - I can't wait to deep dive more. Current structure of the BTree of yours really makes the BTWait a powerful node in general. Since they use coroutines, I feel it's much more cleaner. But what if you want create a timer like this, it wouldn't work like I am thinking yes? 

```swift
func _tick(actor, blackboard) -> bool:
  timer += blackboard.get('delta')
  if timer >= max_time:
     timer = 0
     return succeed()
  return fail()
```

**Reply by kagenash1** [reply]
This would work perfectly as well, except it may not always be the actual time if you're not ticking it at every frame.
If you want to do this kind of thing, I suggest you look into BTGuard. 
They are decorators which allow you to "lock" the child (or an entire sub-tree) for a given time without suspending the execution of the tree. When a guard is locked, it will fail. 
The guard has a lock_if property, which means lock if the child succeeds/fails. 

They also allow you to lock and unlock them when any other node in the tree is ticked based on the tick result.
If you don't specify a locker, the guard will lock based on the children. If you don't specify an unlocker, the guard will unlock based on the lock time, which is ignored if you have an unlocker.

This is a quick example for a simple pathfinding AI. In this case I wanted to calculate the path only once, and recalculate it only after the player is reached and becomes unreached again (to avoid 60 pathfinding calculations per second). If the player is reached I put a wait 1 sec as a placeholder.


![Screenshot 2021-03-20 071219](https://user-images.githubusercontent.com/58669655/111861056-a349b000-894b-11eb-9281-871b2e721894.png)

![Screenshot 2021-03-20 071023](https://user-images.githubusercontent.com/58669655/111861021-5b2a8d80-894b-11eb-8af4-7c0634ac5694.png)

**Reply by kagenash1** [reply]
This, for instance, will do what you want :)
Here the unlock_if property is not taken into account cause no unlocker is specified, so it checks the lock_time (which can be anything you want, in seconds)

![Screenshot 2021-03-20 072014](https://user-images.githubusercontent.com/58669655/111861200-bd37c280-894c-11eb-9ff5-3bcca0f4c09c.png)

**Reply by ashfidable** [reply]
Man, you are the best! It's working perfectly on my end. I am loving the BTGuard! I am using it for my low-reactive enemies now and it was just very intuitive. Really loving the fact that Locker and Unlocker can actually take anything from the BTree. Talk about designer friendly, you are simply nailing it. 

I don't want to bother you no more, but I put the behavior tree on other project and the tree is 101 same no difference but sadly it's saying error at 
```swift
assert(get_child_count() > 1)
``` 
for BTComposite Node. Here's the tree - I am sure I am missing out something right here - It's totally fine if you are busy, I am sure I will figure it out.
![tree](https://user-images.githubusercontent.com/41086166/111861926-f4b16980-8962-11eb-9bc0-eab0f98442ec.png)

**Reply by kagenash1** [reply]
Basically composite nodes are supposed to take more than 1 node. That's a limit I put to let you see when you misuse a component (and also avoid useless ticking which might affect performance)
What is the purpose of the first sequence? It only has one child, so there is no point in having it :)

Btw, thanks to you I'm realizing a lot of improvements I can make so feel free to spam as many questions as you want! For example I'm thinking I should print out some errors as well when doing thise assertions

**Reply by ashfidable** [reply]
Oh, that makes sense. Initially, I had another sequence branch. So once that sequence is done which is under the main sequence branch -> go to the other sequence branch. This whole thing didn't give me on the other project though. These are just for checking ofcourse. 

EDIT: How would you stack sequences then for the above question ^

Also definitely man! You don't know how much you are helping me here - thanks for replying to everything. Godot nodes does make it easier to see all the shortcomings but honestly so far it's been really robust asset and if there are problems, it's on my end.

**Reply by kagenash1** [reply]
Look carefully: the first sequence only has one child (the second sequence). So either you add another child or remove it, because it doesn't do anything now. You could just take the second sequence and put it directly under the tree and remove the first one.

**Reply by ashfidable** [reply]
So basically the thing I wanted to do was:
1. Sequence to do successful tasks in order: Patrol 2 points Sequence (if successful then) -> Another sequence of him just talking around (for now, I will put few timers to see this but if successful) -> Do Something else
2. Actually that's all. I want them to do it successfully in order to go to next one, that's fair that they have selectors inside few sequences but that's working fine.

Well this is close example of multiple sequences really, but instead of application main root which is a selector - it's main sequence. Mind you, this works perfectly fine if I remove the assert line. But it makes sense to have one composite node at the root of the tree - but I am not sure if someone wants to chain them inside a sequence? 
![A-behavior-tree-example-for-the-patrol-application](https://user-images.githubusercontent.com/41086166/111862887-2f1e0500-8969-11eb-99db-ca5f44927821.png)

**Reply by kagenash1** [reply]
The problem here is that your first sequence only has one child. Of course you can recursively stack sequences, but you have to make sure a sequence has more than one child (otherwise it serves no purpose).

**Reply by kagenash1** [reply]
As you can see this sequence only has one child. 
It will work fine if you add another one (or if you remove the assert line lol. That is just there to help you avoid having useless sequences but things will still work if you remove it)

![Screenshot 2021-03-20 084748](https://user-images.githubusercontent.com/58669655/111862982-faa24d00-8958-11eb-9e61-99613aca83fa.png)

**Reply by ashfidable** [reply]
![tree](https://user-images.githubusercontent.com/41086166/111862985-c2573a80-8969-11eb-9320-61bcc2086174.png)
So this should be working right? But this does give the Assertion failed error because the child count is above 1 right?

**Reply by kagenash1** [reply]
No because BTSequence2 has one child :) Same problem different node. A sequence must have more than 1 child to be a sequence.
assert(get_child_count() > 1) means "Make sure the child count IS above one"

**Reply by ashfidable** [reply]
OOF. I am the dumbest person alive. Here I was thinking assertion crashes when the condition IS met. I tried putting two timers and it's working. God, I feel embarrassed to keep making this mistake. I did change the assert line to print line to remind me what's wrong in it.

**Reply by kagenash1** [reply]
😆 don't worry. It's clear I should print some kind of error in the next version.
The assertion is there to prevent doing meaningless things, but obviously if you remove it it will work. 
My suggestion is you not remove it because I think it helps: I very often found myself having unnecessarily big trees because I accidentally left single-child sequences or selectors. 
What you could (and should) do if you just wanna lay out an abstract version of the tree, without implementing the actions, is to populate the sequence with empty BTLeaf nodes (instead of wait nodes which will stop the tree), and give them proper names, but without extending the script to implement the action. Empty BTLeaf will work fine and they will just succeed() at every tick.

**Reply by ashfidable** [reply]
Woah, that's actually pretty good suggestion! I am already saving branches as scenes (God, Godot is freaking amazing at that) and this will make less mistakes on my part. 

I don't know how to thank you for helping me with this. I am getting better at this. Seriously, go to asset for any game with complex AI to me. There are other behavior trees but the discussion is too minimal and there are not really much of starter BTNodes to do whatever you want. This one got almost everything I want from nodes. 

Coming to think of it, when do you see 1.2.0 coming since it's really huge in update compared to incremental ones?

**Reply by kagenash1** [reply]
Yeah that's good! I tend to have several trees rather than a big one. Then I can separate my AI more, for example if I have an enemy I will make a BT for pathfinding, one for movement, one for shooting, and so on. They are more controllable this way, and the good thing is they can all use the same blackboard to communicate between each other.

I'm glad you are comfortable with my tool, I hope you can achieve the AI you want and break the monotony of game swith state machine AI, lol. Looking forward to what you create.

As for the new version, I don't know really. I am developing the package as I progress with my own game. The next version will be 1.1.3, mostly with bug fixes and minor improvements, and that will probably be the stable version for the time being. But I am not rushing it cause I wanna make sure I address every bug, so as long as I find stuff to fix I'm not releasing it. 
Then 1.2.0 will start being worked on but that will require even more time, and since 1.1.3 will be stable development won't be so urgent, cause I also wanna focus on my game and that version will be good enough.

To summarize, I'm glad you are looking forward to it but don't overthink it, 1.1.3 will be a very "light" update and you will have no problems using it with your previous trees. So have fun with the current version, cause 1.2.0 will take quite the time 😆 

(This is unless some kind of driving force *cough* *cough* donations *cough* turns up, or the community grows and someone happens to develop good enough features for the new version)

**Reply by ashfidable** [reply]
That's great! I am planning to wrap BT with one or two states for aware and unaware state. So far, it's working pretty good! 

Please take your time with 1.2.0, it's a really huge update with graphedit and everything. I am really happy to know that 1.1.3 will not touch the logic of our action and decorator codes! 

Keep doing what you do man, one of the best packages in recent times for sure on godot.
